File: programmers/cpp/programmers_heap_1.cpp
#include <string>
#include <vector>
#include <queue>
#include <functional>

using namespace std;

int solution(vector<int> scoville, int K) {
	int answer = 0;
	priority_queue<int, vector<int>, greater<int>> q;

	for (int i = 0; i < scoville.size(); i++)
		q.push(scoville[i]);
	while (1)
	{
		if (q.size() == 1 && q.top() < K)
			return -1;
		else if (q.size() == 1)
			return answer;
		if (q.top() >= K)
			return answer;
		int temp1 = q.top();
		q.pop();
		int temp2 = q.top();
		q.pop();
		q.push(temp1 + (temp2 * 2));
		answer++;
	}

	return answer;
}

File: programmers/python/programmers_heap_1.py
import heapq

def solution(scoville, K):
	heapq.heapify(scoville)
	answer = 0
	while True:
		if scoville[0] > K:
			break
		elif len(scoville) == 1 and scoville[0] < K:
			answer = -1
			break
		sco1 = heapq.heappop(scoville)
		sco2 = heapq.heappop(scoville)
		heapq.heappush(scoville, sco1 + (sco2 * 2))
		answer += 1
	return answer